pkg/client/useragent.go

```diff
@@ -0,0 +1,65 @@
+package client
+
+import (
+	"fmt"
+
+	"github.com/rancher/lasso/pkg/log"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/rest"
+)
+
+type sharedClientFactoryWithAgent struct {
+	SharedClientFactory
+
+	userAgent string
+}
+
+// NewSharedClientFactoryWithAgent returns a sharedClientFactory that is equivalent to client factory with the addition
+// that it will augment returned clients from ForKind, ForResource, and ForResourceKind with the passed user agent.
+func NewSharedClientFactoryWithAgent(userAgent string, clientFactory SharedClientFactory) SharedClientFactory {
+	return &sharedClientFactoryWithAgent{
+		SharedClientFactory: clientFactory,
+		userAgent:           userAgent,
+	}
+}
+
+func (s *sharedClientFactoryWithAgent) ForKind(gvk schema.GroupVersionKind) (*Client, error) {
+	client, err := s.SharedClientFactory.ForKind(gvk)
+	if err != nil {
+		return client, err
+	}
+	return s.clientWithAgent(client)
+}
+
+func (s *sharedClientFactoryWithAgent) ForResource(gvr schema.GroupVersionResource, namespaced bool) (*Client, error) {
+	client, err := s.SharedClientFactory.ForResource(gvr, namespaced)
+	if err != nil {
+		return client, err
+	}
+	return s.clientWithAgent(client)
+}
+
+func (s *sharedClientFactoryWithAgent) ForResourceKind(gvr schema.GroupVersionResource, kind string, namespaced bool) *Client {
+	client := s.SharedClientFactory.ForResourceKind(gvr, kind, namespaced)
+
+	clientWithAgent, err := s.clientWithAgent(client)
+	if err != nil {
+		log.Debugf(fmt.Sprintf("Failed to return client with agent [%s]: %v", s.userAgent, err))
+	}
+	return clientWithAgent
+}
+
+func (s *sharedClientFactoryWithAgent) clientWithAgent(client *Client) (*Client, error) {
```

Comment:
Why not use `client.WithAgent`?

Reply:
🤔 good call, WithAgent properly creates a copy as well. Replaced.